ratelimit/redis.go

```diff
@@ -0,0 +1,166 @@
+package ratelimit
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/go-redis/redis"
+	log "github.com/sirupsen/logrus"
+)
+
+// RedisOptions is used to configure the redis.Ring
+type RedisOptions struct {
+	// Addrs are the list of redis shards
+	Addrs []string
+}
+
+// clusterLimitRedis stores all data required for the cluster ratelimit.
+type clusterLimitRedis struct {
+	group   string
+	maxHits int64
+	window  time.Duration
+	ring    *redis.Ring
+}
+
+// newClusterRateLimiterRedis creates a new clusterLimitRedis for given
+// Settings. Group is used to identify the ratelimit instance, is used
+// in log messages and has to be the same in all skipper instances.
+func newClusterRateLimiterRedis(s Settings, options *RedisOptions, group string) *clusterLimitRedis {
+	if options == nil {
+		return nil
+	}
+
+	ringOptions := &redis.RingOptions{
+		Addrs: map[string]string{},
+	}
+	for idx, addr := range options.Addrs {
+		ringOptions.Addrs[fmt.Sprintf("server%d", idx)] = addr
+	}
+
+	ring := redis.NewRing(ringOptions)
+	// TODO(sszuecs): maybe wrap with context and expose a flag
+	//ring = ring.WithContext(context.Background())
+
+	rl := &clusterLimitRedis{
+		group:   group,
+		maxHits: int64(s.MaxHits),
+		window:  s.TimeWindow,
+		ring:    ring,
+	}
+
+	var err error
+
+	err = backoff.Retry(func() error {
+		_, err = rl.ring.Ping().Result()
+		if err != nil {
+			log.Infof("Failed to ping redis, retry with backoff: %v", err)
+		}
+		return err
+	}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), 7))
+
+	if err != nil {
+		log.Errorf("Failed to connect to redis: %v", err)
+		return nil
+	}
+	log.Debug("Redis ring is reachable")
+
+	return rl
+}
+
+// Allow returns true if the request calculated across the cluster of
+// skippers should be allowed else false. It will share it's own data
+// and use the current cluster information to calculate global rates
+// to decide to allow or not.
+func (c *clusterLimitRedis) Allow(s string) bool {
```

Comment:
It would be great to know the two `redis`-round-trips overhead here. Some of the performance tests show, that it scales bad with a high number of requests and, that there is a huge potential for improving the algorithm by being smarter in checking and denying requests. The algorithm is still very simple so some tests to improve the performance will not hurt. I have some ideas how to do this we could talk about.

Reply:
done

Reply:
The simple improvement is to store the last `RetryAfter()` value to block any request until this time is reached before checking the value in redis. This would at least reduce the overhead for *high volume overload* scenarios and allow proper rejection in such situations.